dev/planning-example.py

```diff
@@ -0,0 +1,265 @@
+from astroplan.entity import Constraints
+
+# times can be passed in as strings (interpreted as for get_date()) or
+# as astropy Time or datetime objects
+cts = Constraints(time_start="2015-05-01 18:30", time_end="2015-05-02 05:30",
+                  alt_min=15.0*u.deg, alt_max=89.0*u.deg, duration=30*u.min)
+
+# define a target
+tgt = SiderealTarget(name='S5', ra='14:20:00.00', dec='48:00:00.00')
+
+# returns an object with information about the observability with these
+# constraints
+info = obs.observable(tgt, cts)
+
+'''
+attributes of return object:
+  observable: a boolean -- whether target is observable at desired constraints
+  rise_time: a datetime -- first time target is visible in at these conditions
+  set_time: a datetime -- time target is no longer visible at these conditions
+
+Use of a return object means that additional items can be added in the
+future without breaking the API (e.g. subclasses are free to attach
+additional items).  It can be also useful to attach an "explanation" for why
+something cannot be observed.
+'''
+
+# same question, but specify the lower elevation bound to be the
+# telescope minimum and I will specify an airmass of 1.2 for the target
+cond = Constraints(time_start="2015-05-01 18:30", time_end="2015-05-02 05:30",
+                   alt_min=15.0*u.deg, alt_max=89.0*u.deg, duration=30*u.min,
+                   airmass=1.2)
+info = obs.observable(tgt, cond)
+
+'''
+Suggest that Constraints implement a time range, an altitude range and an
+airmass limit at beginning.  Other things that could be added later:
+ - azimuth range
+ - airmass range
+
+Idea is that Constraints can be subclassed to add additional constraints.
+'''
+
+# ======================================================
+# Other useful calculations wrt an observer and a target
+#=======================================================
+
+# calculate the distance in alt and az degrees between two targets at
+# the given time (e.g. to calculate slew time)
+sf = SiderealTarget(name='Sf', ra='09:40:00.00', dec='43:00:00.00')
+sm = SiderealTarget(name='Sm', ra='10:30:00.00', dec='36:00:00.00')
+t = obs.get_date("2015-05-01 23:00")
+
+# returns a vector or other astropy quantity specifying a delta in alt, az
+obs.distance(sf, sm, t)
+
+# tell me about object sm in relation to observer obs at time t
+# returns a CalculationResult object
+cr = obs.calc(sm, t)
+
+"""
+Note use of CalculationResult object--most computations are lazily delayed
+until they are requested from the object.  This saves a lot of time if many
+objects are evaluated wrt a given time and observation position and we only
+need some of the calculation results (e.g. for scheduling).  For example, we
+might evaluate 400 targets against a given time slot and we are only
+interested in their altitude and airmass--calculation time dominates when
+scheduling from a large number of potential targets.
+
+CalculationResult objects contain a number of properties, most of which
```

Comment:
This comes again to the convenience factor.  If I am an observer doing some simple scripting, I'd much rather write:
r = calc(sm, m, t)
r.airmass
r.pang
r.ha

than:
am = airmass(obs, sm, t)
pang = parallactic_angle(obs, sm, t)
ha = hour_angle(obs, sm, t)

Again, for users using the api to write layers of software on top this is not a big issue, since they will write their own convenience methods.  I guess this raises the question of for what range of users are we are targeting this package?  My feeling was that the package could be used interactively or in an ipython notebook for simple observation planning as well as for writing sophisticated layers on top.  If we want to include the former users then we need to provide convenient _shorthand_ ways of expressing things or else we leave it to someone else to write that layer.